Issue template form

redstonedesigner opened the pull request:
This pull request removes the old file, located at .github/issue_template.md (this is a legacy method of defining issue templates).
The new method of creating an issue template also allows for forms to be created when users are submitting an issue.  When creating the issue template, I have used the current issue template as closely as possible.
This pull request also disables blank issue creation, meaning that all new issues must use the form(s) configured to submit an issue.

This form can be tested by going to https://github.com/redstonedesigner/gitea-issue-form-testing and creating a new issue.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+    label: Git Version
+    description: The version of git running on the server's systemm
+  validations:
+    required: true
```
I don't think the git version should be required.
Until now, most of the issues I opened could be managed without knowing the git version. Additionally, doing it that way disables sending bugs present on try.gitea.io as I have no idea what git version is running there.

redstonedesigner replied:
I see your point there, in my next commit this will be patched.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+    label: Operating System
+    description: The operating system you are using to run Gitea
+  validations:
+    required: true
```
I'd also say that the OS is not required, at  least for issues that deal with the frontend.

redstonedesigner replied:
Agreed.  I'll disable the requirement in my next commit.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+      Please also tell us how you are running gitea, e.g. if it is being run from docker, a command-line, systemd etc.
+      If you are using a package or systemd tell us what distribution you are using
+  validations:
+    required: true
```
Also not always required. What should I say about try.gitea.io?

redstonedesigner replied:
As far as I am aware, try.gitea.io is a testing environment only.  There is a section below regarding using try.gitea.io to test things that you find on your own installations.
This is more checking wheter they are using https://gitea.com, another hosted environment or their own self hosted variation.
I can remove the required flag from this input, however I bleieve that it would be valuable information for users to provide in terms of where they are using Gitea.

delvh replied:
Yeah, I know.
But the comfort of try.gitea.com is exactly that the latest features can already be tested. And sometimes, there are bugs in these new features. Then, you can refer to try.gitea.com for these bugs.
That's what I mean by that.

redstonedesigner replied:
To the best of my knowledge, try.gitea.io is a testing environment only and should only be used for reproducing bugs and trying out gitea.
I think that this would be valuable information for contributors and collaborators to have, so instead of disabling the requirement I might add something that prompts the person to write if they are using try.gitea.io .
Thoughts?

delvh replied:
Okay, sounds fine.
(Since I am most likely affected the most by this - I'd say that I am the person with the highest count of issues opened from testing on try.gitea.com) that sounds reasonable.
Yeah, I can support that.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+    - MSSQL
+    - SQLite
+  validations:
+    required: true
```
Once again: try.gitea.io. Should not be required.
Also, not required for pure frontend issues.

redstonedesigner replied:
Removed in next commit.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+    - "Yes"
+    - "No"
+  validations:
+    required: true
```
Okay, here it is a little different. I can see why you would require it here.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+    description: Gitea version (or commit reference) your instance is running
+    placeholder: v1.14.3
+  validations:
+    required: true
```
Yes, here it can be required.

delvh reviewed (commented):
Apart from the mandatory requirements: Well done :+1:

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,96 @@
+  attributes:
+    label: Gitea Version
+    description: Gitea version (or commit reference) your instance is running
+    placeholder: v1.14.3
```
Is there maybe a way so that the placeholder will also be used as default?
Because currently, it is only displayed when no text is present.
GitHub still requires you to fill that field before you can submit the issue.

redstonedesigner replied:
I can indeed specify a default value, however this would need to be updated every release.

delvh replied:
But the placeholder not?

redstonedesigner replied:
The placeholder could also be updated, however I consider placeholders as example values and not the value that a user should enter.

redstonedesigner commented:
Thank you for the feedback @delvh - I believe I have addressed your comments in my most recent commit.  The test repository has also been updated with the latest version of this template.

Please let me know if you have any further input.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -21,28 +21,25 @@ body:
     label: How are you running Gitea?
     description: |
-      Please include information on whether you built gitea yourself, used one of our downloads or are using some other package
+      Please include information on whether you built gitea yourself, used one of our downloads, are using https://try.gitea.io or are using some other package
```
:+1:

redstonedesigner commented:
Just had a thought - Would there be any further issue types that would require more specific information or additional information?

With the new issue template system, it is possible to customise the experience for the user depending on the type of error they are experiencing.  It is also possible to use labels in the templates, which would be added depending on the form that's submitted.

delvh commented:
Yes, while that is definitely possible and should be implemented sometime, I think that's beyond the scope of this PR.
Unless of course you do want to do it now.

redstonedesigner commented:
If we're changing the workflow for creating issues in this PR, I feel like it might be worth making all the changes in this PR to avoid multiple PRs for a similar issue.
I'm happy to do it now, just let me know what other issue forms you would like added and I can get them configured.

delvh commented:
Well okay, then I would first recommend a second template for Feature Requests.
This would be pretty basic and most likely only contain one or two sections: a) the description of this feature (this can be required) and b) an area to show optional screenshots (i.e. of the GitHub implementation or a proposed design)

redstonedesigner commented:
I've gone ahead and put the feature request form onto the testing repository as well as this PR.

delvh commented on .github/ISSUE_TEMPLATE/issue.yaml:
```diff
@@ -0,0 +1,91 @@
+name: New Issue
```
Then this issue type would most likely be `Bug Report`, or something similar…

redstonedesigner replied:
Should now be resolved in the latest commit.  I also changed the file name and description to be more in line with a bug report.

a1012112796 commented:
![image](https://user-images.githubusercontent.com/25342410/124607687-a9cf4980-dea0-11eb-9168-162b086e2eea.png)


How to provide link to try.gitea.io ?

redstonedesigner commented:
> ![image](https://user-images.githubusercontent.com/25342410/124607687-a9cf4980-dea0-11eb-9168-162b086e2eea.png)
> 
> How to provide link to try.gitea.io ?

You're prompted to include it in the Description field, which is below the Logs field.

a1012112796 commented on .github/ISSUE_TEMPLATE/config.yaml:
```diff
@@ -0,0 +1 @@
+blank_issues_enabled: false
```
```suggestion
blank_issues_enabled: true
```

Suggest enable it.

redstonedesigner replied:
If this were to be enabled, people would have the ability to create blank issues without providing any formatting or guidance at all.

a1012112796 replied:
Sometimes it's usefull, such as create a summary or a vote.

delvh commented:
Also, it might be nice to setup the following in the config.yml:
```yaml
contact_links:
  - name: Security issues
    url/mail: security@gitea.io
    about: If your issue is security related, please write a mail to security@gitea.io instead of opening an issue
```
Or something similar, so that the user does not even get the idea to open a public security related issue.
However, I do not know whether GitHub supports mail addresses here, or whether they get problematic. You'll have to try whether that can actually be done, or whether that is yet impossible.

redstonedesigner commented:
> Also, it might be nice to setup the following in the config.yml:
> 
> ```yaml
> contact_links:
>   - name: Security issues
>     url/mail: security@gitea.io
>     about: If your issue is security related, please write a mail to security@gitea.io instead of opening an issue
> ```
> 
> Or something similar, so that the user does not even get the idea to open a public security related issue.
> However, I do not know whether GitHub supports mail addresses here, or whether they get problematic. You'll have to try whether that can actually be done, or whether that is yet impossible.

I would have to use `url: mailto:security@gitea.io`, which I believe might be problematic in terms of rendering.  I will give it a go on the testing repository and determine how it works.

redstonedesigner commented:
It appears there's a URL parser that validates that the URL is indeed a web URL (e.g. prefixed with `https://`).

delvh commented:
That's a pity. This approach would have been much more intuitive than hiding that information inside the bug report as a comment… Who knows, maybe one day GitHub will add this as well?

redstonedesigner commented:
It is less than ideal.  However I'm hoping that by positioning the email address at the very top of the bug report form should get their attention.

redstonedesigner commented:
@delvh Is it worth me tagging all issues created with the Bug Report form with `kind/bug`?
I'm hesitant to do that with `kind/feature` as some feature requests could be considered `kind/enhancement`.

jpraet commented:
Ah, I had been working on this as well: https://github.com/jpraet/issue_template/issues/new/choose

Maybe you could incorporate my config.yml into this PR? I got the mailto security@gitea.io working by using an URL shortener.

One thing I find a little annoying about the issue forms is that fields without value get rendered in the output as well, see https://github.com/github/feedback/discussions/4491.

jpraet commented on .github/ISSUE_TEMPLATE/bug-report.yaml:
```diff
@@ -0,0 +1,91 @@
+    label: Gitea Version
+    description: Gitea version (or commit reference) your instance is running
+    placeholder: v1.14.3
+    value: v1.14.3
```
I don't think we should have a default value. And maybe not a placeholder either. I fear users will leave the default value unchanged, without confirming they are actually using that version.

delvh replied:
Okay, I can live with both. I would have liked it, but I can also see your concern…

redstonedesigner replied:
I removed the placeholder and default version in my recent commit.

redstonedesigner commented:
> Maybe you could incorporate my config.yml into this PR? I got the mailto [security@gitea.io](mailto:security@gitea.io) working by using an URL shortener.

I hadn't considered that, but I've incorporated your conifg.yml into the PR as you've suggested - I wasn't even aware most of those links existed.  If there are any other links that should be added, please do let me know.
 
> One thing I find a little annoying about the issue forms is that fields without value get rendered in the output as well, see [github/feedback#4491](https://github.com/github/feedback/discussions/4491).

That is a little frustrating, but hopefully it gets resolved soon.  At the very least, it helps people understand what form was submitted.

redstonedesigner commented:
It appears as though I failed some unit tests wihtout actually making any changes to the source code of the project?

jpraet commented on .github/ISSUE_TEMPLATE/bug-report.yaml:
```diff
@@ -0,0 +1,89 @@
+- type: markdown
+  attributes:
+    value: |
+      NOTE: If your issue is a security concern, please send an email to security@gitea.io instead of opening a public issue.
```
Should this still be here now that we have it already on the new issue landing page with the config.yaml?
Maybe an extra reminder here is not that bad.

redstonedesigner replied:
Too many reminders is never a bad thing if it stops security vulnerabilities being made public.
Ideally, it will make people reporting bugs think more than once, and that's if they even read far enough down.

jpraet commented on .github/ISSUE_TEMPLATE/bug-report.yaml:
```diff
@@ -0,0 +1,89 @@
+  attributes:
+    value: |
+      1. Please speak English, this is the language all maintainers can speak and write.
+      2. Please ask questions or configuration/deploy problems on our Discord
```
Should this still be here now that we have it already on the new issue landing page with the config.yaml?

redstonedesigner replied:
Always useful to have this information in multiple locations.  It also keeps it on the page where they are submitting the form.

jpraet commented on .github/ISSUE_TEMPLATE/bug-report.yaml:
```diff
@@ -0,0 +1,89 @@
+      2. Please ask questions or configuration/deploy problems on our Discord
+         server (https://discord.gg/gitea) or forum (https://discourse.gitea.io).
+      3. Please take a moment to check that your issue doesn't already exist.
+      4. Make sure it's not mentioned in the FAQ (https://docs.gitea.io/en-us/faq)
```
Should this still be here now that we have it already on the new issue landing page with the config.yaml?

redstonedesigner replied:
Again, keeps it present and current so that they can see it.  It never hurts to remind them multiple times.

jpraet commented on .github/ISSUE_TEMPLATE/bug-report.yaml:
```diff
@@ -0,0 +1,89 @@
+  id: screenshots
+  attributes:
+    label: Screenshots
+    description: If this issue involves the Web Interface, please provide a screenshot or multiple screenshots
```
Should we add an input field asking for the browser version for issues that involve the web interface?

delvh replied:
Now that I think about it, it might not be a bad idea to completely separate issues involving web interface bugs from the rest of the bugs…
The report requirements might differ too much.

redstonedesigner replied:
If we would like to do that, I can definitely create a new form.

redstonedesigner replied:
This has been added in the most recent commit and has also been added to the testing repository - https://github.com/redstonedesigner/gitea-issue-form-testing

jpraet commented on .github/ISSUE_TEMPLATE/feature-request.yaml:
```diff
@@ -0,0 +1,16 @@
+name: Feature Request
+description: Got an idea for a feature that Gitea doesn't have currently?  Submit your idea here!
+body:
```
We should also add some notes here about using the English language, and about checking first that the feature hasn't already been requested.

redstonedesigner replied:
We don't have much space here, but I could add a markdown system similar to the bug report to the feature template.

redstonedesigner replied:
I've gone ahead and added a message to the top of the feature request template.

techknowlogick commented on .github/issue_template.md:
```diff
@@ -1,42 +0,0 @@
-<!-- NOTE: If your issue is a security concern, please send an email to security@gitea.io instead of opening a public issue -->
```
Thanks for this PR :)

Rather than removing this file, could you instead move it to the `.gitea` directory (which doesn't yet exist), so that when we move to gitea.com if issue templates don't yet exist as a feature in Gitea then we still have this as a fallback

redstonedesigner replied:
I can definitely do that.  I wasn't aware that there is a `.gitea` directory, and I will add the file accordingly.

redstonedesigner commented:
Does anyone have any further input on this?
